File: test/core/seed_accessor_test.cpp
/*
 * Copyright 2017 Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "core/seed_accessor.hpp"

#include <gtest/gtest.h>

#include <memory>

#include "core/colonio.pb.h"
#include "core/node_id.hpp"
#include "core/packet.hpp"
#include "core/random.hpp"
#include "core/scheduler.hpp"
#include "test_utils/all.hpp"

using namespace colonio;

const std::string SEED_URL("https://localhost:8080/test");

class SeedAccessorDelegateStub : public SeedAccessorDelegate {
  std::mutex mtx;
  bool _on_change_state;
  bool _on_error;
  bool _on_recv_config;
  std::unique_ptr<const Packet> last_packet;
  bool _on_require_random;

 public:
  SeedAccessorDelegateStub() {
    reset();
  }

  void reset() {
    _on_change_state = false;
    _on_error        = false;
    _on_recv_config  = false;
    last_packet.reset();
    _on_require_random = false;
  }

  bool called_on_change_state() {
    std::lock_guard<std::mutex> lock(mtx);
    return _on_change_state;
  }

  bool called_on_error() {
    std::lock_guard<std::mutex> lock(mtx);
    return _on_error;
  }

  bool called_on_recv_config() {
    std::lock_guard<std::mutex> lock(mtx);
    return _on_recv_config;
  }

  bool called_on_recv_packet() {
    std::lock_guard<std::mutex> lock(mtx);
    return static_cast<bool>(last_packet);
  }

  const Packet* packet() {
    std::lock_guard<std::mutex> lock(mtx);
    return last_packet.get();
  }

  void seed_accessor_on_change_state() override {
    std::lock_guard<std::mutex> lock(mtx);
    _on_change_state = true;
  }

  void seed_accessor_on_error(const std::string& message) override {
    std::lock_guard<std::mutex> lock(mtx);
    _on_error = true;
  }

  void seed_accessor_on_recv_config(const picojson::object& config) override {
    std::lock_guard<std::mutex> lock(mtx);
    _on_recv_config = true;
  }

  void seed_accessor_on_recv_packet(std::unique_ptr<const Packet> packet) override {
    std::lock_guard<std::mutex> lock(mtx);
    last_packet = std::move(packet);
  }

  void seed_accessor_on_require_random() override {
    std::lock_guard<std::mutex> lock(mtx);
    _on_require_random = true;
  }
};

TEST(SeedAccessorTest, single_node) {
  AsyncHelper helper;
  TestSeed seed;
  seed.run();

  Random random;
  std::unique_ptr<Scheduler> scheduler(Scheduler::new_instance(logger));

  SeedAccessorDelegateStub delegate;
  SeedAccessor accessor(
      logger, *scheduler, NodeID::make_random(random), delegate, SEED_URL, std::string(""), 30 * 1000, true);

  scheduler->add_task(nullptr, [&]() {
    accessor.enable_polling(true);
    helper.pass_signal("polling");
  });
  helper.wait_signal("polling");
  EXPECT_TRUE(delegate.called_on_change_state());
  EXPECT_EQ(accessor.get_auth_status(), AuthStatus::NONE);
  EXPECT_EQ(accessor.get_link_state(), LinkState::CONNECTING);
  EXPECT_FALSE(delegate.called_on_error());
  EXPECT_FALSE(accessor.last_request_had_error());

  // wait for authenticate
  EXPECT_TRUE(helper.eventually(
      [&]() {
        // call `called_on_change_state` for pass mutex
        return delegate.called_on_change_state() && accessor.get_link_state() == LinkState::ONLINE;
      },
      300 * 1000, 100));
  EXPECT_EQ(accessor.get_auth_status(), AuthStatus::SUCCESS);
  EXPECT_TRUE(delegate.called_on_recv_config());
  EXPECT_TRUE(accessor.is_only_one());
  EXPECT_FALSE(accessor.last_request_had_error());

  // wait for disconnect
  delegate.reset();
  scheduler->add_task(nullptr, [&]() {
    accessor.disconnect();
  });

  EXPECT_TRUE(helper.eventually(
      [&]() {
        return delegate.called_on_change_state() && accessor.get_link_state() == LinkState::OFFLINE;
      },
      300 * 1000, 100));
  EXPECT_EQ(accessor.get_auth_status(), AuthStatus::SUCCESS);
  EXPECT_FALSE(accessor.last_request_had_error());
}

TEST(SeedAccessorTest, relay_poll) {
  AsyncHelper helper;
  TestSeed seed;
  seed.run();

  Random random;
  std::unique_ptr<Scheduler> scheduler(Scheduler::new_instance(logger));

  SeedAccessorDelegateStub delegate1;
  SeedAccessor accessor1(
      logger, *scheduler, NodeID::make_random(random), delegate1, SEED_URL, std::string(""), 30 * 1000, true);
  accessor1.tell_online_state(true);

  SeedAccessorDelegateStub delegate2;
  SeedAccessor accessor2(
      logger, *scheduler, NodeID::make_random(random), delegate2, SEED_URL, std::string(""), 30 * 1000, true);
  accessor2.tell_online_state(true);

  scheduler->add_task(nullptr, [&]() {
    accessor1.enable_polling(true);
    accessor2.enable_polling(true);
  });

  // wait for both of accessor will be online
  EXPECT_TRUE(helper.eventually(
      [&]() {
        return delegate1.called_on_change_state() && accessor1.get_link_state() == LinkState::ONLINE &&
               !accessor1.is_only_one() && delegate2.called_on_change_state() &&
               accessor2.get_link_state() == LinkState::ONLINE && !accessor2.is_only_one();
      },
      300 * 1000, 100));

  // relay a packet from accessor1
  NodeID dst_nid = NodeID::make_random(random);
  NodeID src_nid = NodeID::make_random(random);
  uint32_t id    = random.generate_u32();
  scheduler->add_task(nullptr, [&]() {
    std::unique_ptr<proto::PacketContent> p = std::make_unique<proto::PacketContent>();
    proto::Error* e                         = p->mutable_error();
    e->set_code(100);
    e->set_message("test");
    std::unique_ptr<Packet> packet = std::make_unique<Packet>(dst_nid, src_nid, id, std::move(p), PacketMode::NO_RETRY);
    accessor1.relay_packet(std::move(packet));
  });

  // accessor2 should receive the packet
  EXPECT_TRUE(helper.eventually(
      [&]() {
        return delegate2.called_on_recv_packet();
      },
      300 * 1000, 100));
  const Packet* recv = delegate2.packet();
  EXPECT_EQ(recv->dst_nid, dst_nid);
  EXPECT_EQ(recv->src_nid, src_nid);
  EXPECT_EQ(recv->id, id);
  EXPECT_EQ(recv->hop_count, uint32_t(0));
  EXPECT_EQ(recv->mode, PacketMode::NO_RETRY);
  EXPECT_EQ(recv->content->as_proto().error().code(), uint32_t(100));
  EXPECT_EQ(recv->content->as_proto().error().message(), "test");
}

TEST(SeedAccessorTest, detect_error) {
  AsyncHelper helper;
  // seed is not running
  // TestSeed seed;
  // seed.run();

  Random random;
  std::unique_ptr<Scheduler> scheduler(Scheduler::new_instance(logger));

  SeedAccessorDelegateStub delegate;
  SeedAccessor accessor(
      logger, *scheduler, NodeID::make_random(random), delegate, SEED_URL, std::string(""), 30 * 1000, true);

  scheduler->add_task(nullptr, [&]() {
    accessor.enable_polling(true);
  });

  // wait for authenticate
  EXPECT_TRUE(helper.eventually(
      [&]() {
        return accessor.last_request_had_error();
      },
      300 * 1000, 100));
}

File: internal/network/seed_accessor_test.go
//go:build !js

/*
 * Copyright 2017- Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network

import (
	"context"
	"fmt"
	"log/slog"
	"slices"
	"sync"
	"testing"
	"time"

	"github.com/llamerada-jp/colonio/config"
	"github.com/llamerada-jp/colonio/internal/proto"
	"github.com/llamerada-jp/colonio/internal/shared"
	"github.com/llamerada-jp/colonio/test/testing_seed"
	testUtil "github.com/llamerada-jp/colonio/test/util"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type authenticator struct {
	expectedAccount map[string][]byte
}

func (auth *authenticator) Authenticate(token []byte, nodeID string) (bool, error) {
	expectedToken, ok := auth.expectedAccount[nodeID]
	if !ok {
		return false, nil
	}
	return slices.Equal(expectedToken, token), nil
}

func TestSeedAccessorAuthSuccess(t *testing.T) {
	nodeID := shared.NewRandomNodeID()

	auth := &authenticator{
		expectedAccount: map[string][]byte{
			nodeID.String(): []byte("token"),
		},
	}

	testSeed := testing_seed.NewTestingSeed(
		testing_seed.WithAuthenticator(auth),
		testing_seed.WithRevision(3.14),
		testing_seed.WithSessionTimeout(35*time.Second),
		testing_seed.WithPollingTimeout(12*time.Second),
	)
	defer testSeed.Stop()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mtx := sync.Mutex{}
	events := ""

	config := &seedAccessorConfig{
		ctx:    ctx,
		logger: slog.Default(),
		transporter: DefaultSeedTransporterFactory(&SeedTransporterOption{
			Verification: false,
		}),
		url:          fmt.Sprintf("https://localhost:%d/test", testSeed.Port()),
		nodeID:       nodeID,
		token:        []byte("token"),
		tripInterval: 10 * time.Second,

		authorizeFailedEventHandler: func() {
			assert.FailNow(t, "authentication should be success")
		},
		changeStateEventHandler: func() {
			mtx.Lock()
			defer mtx.Unlock()
			events += "s"
		},
		recvConfigEventHandler: func(conf *config.Cluster) {
			mtx.Lock()
			defer mtx.Unlock()
			events += "c"

			assert.Equal(t, 3.14, conf.Revision)
			assert.Equal(t, 35*time.Second, conf.SessionTimeout)
			assert.Equal(t, 12*time.Second, conf.PollingTimeout)
		},
		recvPacketEventHandler: func(_ *shared.Packet) {
			assert.FailNow(t, "no one should send packet on this test")
		},
		requireRandomConnectEventHandler: func() {
			assert.FailNow(t, "random connect event should not be happened when only one node")
		},
	}

	accessor := newSeedAccessor(config)
	// defer accessor.destruct()
	assert.Eventually(t, func() bool {
		return accessor.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	mtx.Lock()
	// change state to online, recv config
	assert.Equal(t, "cs", events)
	mtx.Unlock()

	accessor.destruct()
	assert.True(t, assert.Eventually(t, func() bool {
		return !accessor.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond))

	mtx.Lock()
	// change state to online, recv config, change state to offline
	assert.Equal(t, "css", events)
	mtx.Unlock()
}

func TestSeedAccessorAuthFailure(t *testing.T) {
	nodeIDs := testUtil.UniqueNodeIDs(2)
	nodeID, differentID := nodeIDs[0], nodeIDs[1]

	auth := &authenticator{
		expectedAccount: map[string][]byte{
			nodeID.String(): []byte("token"),
		},
	}

	testSeed := testing_seed.NewTestingSeed(testing_seed.WithAuthenticator(auth))
	defer testSeed.Stop()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mtx := sync.Mutex{}
	events := ""

	config := &seedAccessorConfig{
		ctx:    ctx,
		logger: slog.Default(),
		transporter: DefaultSeedTransporterFactory(&SeedTransporterOption{
			Verification: false,
		}),
		url:          fmt.Sprintf("https://localhost:%d/test", testSeed.Port()),
		tripInterval: 10 * time.Second,

		authorizeFailedEventHandler: func() {
			mtx.Lock()
			defer mtx.Unlock()
			events += "f"
		},
		changeStateEventHandler: func() {
			assert.FailNow(t, "state event should not be happened when authentication failed")
		},
		recvConfigEventHandler: func(_ *config.Cluster) {
			assert.FailNow(t, "config data should not be received when authentication failed")
		},
		recvPacketEventHandler: func(_ *shared.Packet) {
			assert.FailNow(t, "no one should send packet on this test")
		},
		requireRandomConnectEventHandler: func() {
			assert.FailNow(t, "random connect event should not be happened when only one node")
		},
	}

	// authentication failed when token is different
	config.nodeID = nodeID
	config.token = []byte("different")
	accessor := newSeedAccessor(config)
	// defer accessor.destruct()
	assert.Eventually(t, func() bool {
		mtx.Lock()
		defer mtx.Unlock()
		return events == "f"
	}, 3*time.Second, 100*time.Millisecond)
	assert.False(t, accessor.isAuthenticated())
	accessor.destruct()

	// authentication failed when nodeID is different
	config.nodeID = differentID
	config.token = []byte("token")
	accessor = newSeedAccessor(config)
	// defer accessor.destruct()
	assert.Eventually(t, func() bool {
		mtx.Lock()
		defer mtx.Unlock()
		return events == "ff"
	}, 3*time.Second, 100*time.Millisecond)
	assert.False(t, accessor.isAuthenticated())
	accessor.destruct()
}

func TestSeedAccessorActivate(t *testing.T) {
	nodeID := shared.NewRandomNodeID()

	auth := &authenticator{
		expectedAccount: map[string][]byte{
			nodeID.String(): []byte("token"),
		},
	}

	testSeed := testing_seed.NewTestingSeed(testing_seed.WithAuthenticator(auth))
	defer testSeed.Stop()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mtx := sync.Mutex{}
	events := ""

	config := &seedAccessorConfig{
		ctx:    ctx,
		logger: slog.Default(),
		transporter: DefaultSeedTransporterFactory(&SeedTransporterOption{
			Verification: false,
		}),
		url:          fmt.Sprintf("https://localhost:%d/test", testSeed.Port()),
		nodeID:       nodeID,
		token:        []byte("token"),
		tripInterval: 10 * time.Second,

		authorizeFailedEventHandler: func() {
			assert.FailNow(t, "authentication should be success")
		},
		changeStateEventHandler: func() {
			mtx.Lock()
			defer mtx.Unlock()
			events += "s"
		},
		recvConfigEventHandler: func(_ *config.Cluster) {
			mtx.Lock()
			defer mtx.Unlock()
			events += "c"
		},
		recvPacketEventHandler: func(_ *shared.Packet) {
			assert.FailNow(t, "no one should send packet on this test")
		},
		requireRandomConnectEventHandler: func() {
			assert.FailNow(t, "random connect event should not be happened when only one node")
		},
	}

	accessor := newSeedAccessor(config)
	// defer accessor.destruct()
	assert.Eventually(t, func() bool {
		return accessor.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	mtx.Lock()
	assert.Equal(t, "cs", events)
	mtx.Unlock()

	accessor.activate(false)
	assert.Eventually(t, func() bool {
		return !accessor.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	mtx.Lock()
	assert.Equal(t, "css", events)
	mtx.Unlock()

	accessor.activate(true)
	assert.Eventually(t, func() bool {
		return accessor.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	mtx.Lock()
	assert.Equal(t, "csscs", events)
	mtx.Unlock()

	accessor.destruct()
	assert.Equal(t, "csscss", events)
}

func TestSeedAccessorSessions(t *testing.T) {
	nodeIDs := testUtil.UniqueNodeIDs(2)

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	testSeed := testing_seed.NewTestingSeed(
		testing_seed.WithPollingTimeout(500*time.Millisecond),
		testing_seed.WithSessionTimeout(1500*time.Millisecond),
	)

	config := &seedAccessorConfig{
		ctx: ctx,
		transporter: DefaultSeedTransporterFactory(&SeedTransporterOption{
			Verification: false,
		}),
		url:          fmt.Sprintf("https://localhost:%d/test", testSeed.Port()),
		tripInterval: 1 * time.Second,

		authorizeFailedEventHandler: func() {
			assert.FailNow(t, "authentication should be success")
		},
		changeStateEventHandler:          func() {},
		recvConfigEventHandler:           func(_ *config.Cluster) {},
		recvPacketEventHandler:           func(_ *shared.Packet) {},
		requireRandomConnectEventHandler: func() {},
	}

	// create pinger
	configPinger := *config
	configPinger.nodeID = nodeIDs[0]
	configPinger.logger = slog.With("node", "pinger")
	pinger := newSeedAccessor(&configPinger)
	defer pinger.destruct()
	assert.Eventually(t, func() bool {
		return pinger.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)
	assert.True(t, pinger.isOnlyOne())

	// create node will be offline gracefully
	configTarget := *config
	configTarget.nodeID = nodeIDs[1]
	configTarget.logger = slog.With("node", "target")
	target := newSeedAccessor(&configTarget)
	// defer target.destruct()
	assert.Eventually(t, func() bool {
		return target.isAuthenticated() && !target.isOnlyOne() && !pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	target.activate(false)
	assert.Eventually(t, func() bool {
		return !target.isAuthenticated() && pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	target.activate(true)
	assert.Eventually(t, func() bool {
		return target.isAuthenticated() && !target.isOnlyOne() && !pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	target.destruct()
	assert.Eventually(t, func() bool {
		return !target.isAuthenticated() && pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	// create node will be offline ungracefully
	ctx2, cancel2 := context.WithCancel(context.Background())
	configTarget.ctx = ctx2
	target = newSeedAccessor(&configTarget)
	defer target.destruct()
	assert.Eventually(t, func() bool {
		return target.isAuthenticated() && !target.isOnlyOne() && !pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	cancel2() // cause ungraceful offline by canceling context
	assert.Eventually(t, func() bool {
		return pinger.isOnlyOne()
	}, 3*time.Second, 100*time.Millisecond)

	// be offline by session timeout
	testSeed.Stop()
	assert.Eventually(t, func() bool {
		return !pinger.isAuthenticated()
	}, 10*time.Second, 100*time.Millisecond)

	// be online again
	testSeed = testing_seed.NewTestingSeed()
	defer testSeed.Stop()
	assert.Eventually(t, func() bool {
		return pinger.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)
}

func TestSeedAccessorRelayPacket(t *testing.T) {
	// receiver1, receiver2, sender, destination of random, source of random
	nodeIDs := testUtil.UniqueNodeIDs(5)

	// test packet base
	packetBase := shared.Packet{
		DstNodeID: nodeIDs[3],
		SrcNodeID: nodeIDs[4],
		HopCount:  2,
		Mode:      shared.PacketModeRelaySeed,
		Content: &proto.PacketContent{
			Content: &proto.PacketContent_Error{
				Error: &proto.Error{
					Code:    2,
					Message: "test",
				},
			},
		},
	}

	testSeed := testing_seed.NewTestingSeed()
	defer testSeed.Stop()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	mtx := sync.Mutex{}

	config := &seedAccessorConfig{
		ctx: ctx,
		transporter: DefaultSeedTransporterFactory(&SeedTransporterOption{
			Verification: false,
		}),
		url:          fmt.Sprintf("https://localhost:%d/test", testSeed.Port()),
		tripInterval: 1 * time.Second,

		authorizeFailedEventHandler: func() {
			assert.FailNow(t, "authentication should be success")
		},
		changeStateEventHandler:          func() {},
		recvConfigEventHandler:           func(_ *config.Cluster) {},
		recvPacketEventHandler:           func(_ *shared.Packet) {},
		requireRandomConnectEventHandler: func() {},
	}

	// create accessor for sender
	configSender := *config
	configSender.nodeID = nodeIDs[2]
	configSender.logger = slog.With("node", "sender")
	configSender.recvPacketEventHandler = func(packet *shared.Packet) {
		assert.FailNow(t, "sender should not receive packet")
	}
	sender := newSeedAccessor(&configSender)
	defer sender.destruct()

	// create accessor for receiver 1
	var received *shared.Packet
	configReceiver1 := *config
	configReceiver1.nodeID = nodeIDs[0]
	configReceiver1.logger = slog.With("node", "receiver1")
	configReceiver1.recvPacketEventHandler = func(packet *shared.Packet) {
		mtx.Lock()
		defer mtx.Unlock()
		received = packet
	}
	receiver1 := newSeedAccessor(&configReceiver1)

	// wait for sender and receiver1 to be online
	assert.Eventually(t, func() bool {
		return sender.isAuthenticated() && receiver1.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	// check packet content is correctly relayed
	packet := packetBase
	packet.ID = 10
	sender.relayPacket(&packet)
	assert.Eventually(t, func() bool {
		mtx.Lock()
		defer mtx.Unlock()
		return received != nil
	}, 3*time.Second, 100*time.Millisecond)
	require.NotNil(t, received)
	assert.Equal(t, nodeIDs[3].String(), received.DstNodeID.String())
	assert.Equal(t, nodeIDs[4].String(), received.SrcNodeID.String())
	assert.Equal(t, uint32(10), received.ID)
	assert.Equal(t, uint32(2), received.HopCount)
	assert.Equal(t, shared.PacketModeRelaySeed, received.Mode)
	assert.Equal(t, uint32(2), received.Content.GetError().Code)
	assert.Equal(t, "test", received.Content.GetError().Message)
	receiver1.destruct()

	// recreate receiver1 and receiver2
	receiveCount1 := 0
	configReceiver1.recvPacketEventHandler = func(packet *shared.Packet) {
		mtx.Lock()
		defer mtx.Unlock()
		receiveCount1 += 1
	}
	receiver1 = newSeedAccessor(&configReceiver1)
	defer receiver1.destruct()

	configReceiver2 := *config
	configReceiver2.nodeID = nodeIDs[1]
	configReceiver2.logger = slog.With("node", "receiver2")
	receiveCount2 := 0
	configReceiver2.recvPacketEventHandler = func(packet *shared.Packet) {
		mtx.Lock()
		defer mtx.Unlock()
		receiveCount2 += 1
	}
	receiver2 := newSeedAccessor(&configReceiver2)
	defer receiver2.destruct()

	assert.Eventually(t, func() bool {
		return receiver1.isAuthenticated() && receiver2.isAuthenticated()
	}, 3*time.Second, 100*time.Millisecond)

	// send 10 packets and check if receiver1 or receiver2 receive all packets
	for i := 0; i < 10; i++ {
		packet := packetBase
		packet.ID = uint32(100 + i)
		sender.relayPacket(&packet)
	}
	assert.Eventually(t, func() bool {
		mtx.Lock()
		defer mtx.Unlock()
		return receiveCount1+receiveCount2 == 10
	}, 3*time.Second, 100*time.Millisecond)

	// disconnect the receiver that received more packets and check if the other receiver receives all packets
	if receiveCount1 > receiveCount2 {
		receiver1.destruct()
	} else {
		receiver2.destruct()
	}
	receiveCount1 = 0
	receiveCount2 = 0

	for i := 0; i < 10; i++ {
		packet := packetBase
		packet.ID = uint32(200 + i)
		sender.relayPacket(&packet)
	}
	assert.Eventually(t, func() bool {
		mtx.Lock()
		defer mtx.Unlock()
		return receiveCount1+receiveCount2 == 10
	}, 3*time.Second, 100*time.Millisecond)
}
